scripts/check-stratum1-servers.py

```diff
@@ -0,0 +1,94 @@
+#!/usr/bin/env python3
+
+import argparse
+import datetime
+import sys
+import urllib.error
+import urllib.request
+import yaml
+
+LAST_SNAPSHOT_FILE = '.cvmfs_last_snapshot'
+
+
+def error(msg):
+    """Print an error message to stderr and exit."""
+    print(msg, file=sys.stderr)
+    sys.exit(1)
+
+
+def find_stratum1_urls(vars_file):
+    """Find all Stratum 1 URLs in a given Ansible YAML vars file that contains the EESSI CVMFS configuration."""
+    try:
+        group_vars = urllib.request.urlopen(vars_file)
+    except:
+        error(f'Cannot read the file that contains the Stratum 1 URLs from {vars_file}!')
+    try:
+        group_vars_yaml = yaml.safe_load(group_vars)
+        urls = group_vars_yaml['eessi_cvmfs_server_urls'][0]['urls']
+    except:
+        error(f'Cannot parse the yaml file from {vars_file}!')
+    return urls
+
+
+def check_out_of_sync(s1_urls, fqrn, max_snapshot_delay=30, max_snapshot_diff=30):
+    """Check if all the Stratum 1 servers are in sync."""
+    errors = []
+    last_snapshots = {}
+    now = datetime.datetime.utcnow()
+    for s1 in s1_urls:
+        # Get a URL for the CVMFS last snapshot json file.
+        s1_json = s1.replace('@fqrn@', fqrn) + '/' + LAST_SNAPSHOT_FILE
+        try:
+            last_snapshot = urllib.request.urlopen(s1_json).read().strip().decode('UTF-8')
```

Comment:
Can/should we also log the revision (or whatever it's called), and maybe just produce a warning in case one of the Stratum-1's is serving an old one? That could happen from time to time, but it should be an exception.

Is there a way in which the snapshot time could be recent, but the revision of the repo being served is horribly old somehow? Perhaps due to configuration issues in S1?

Reply:
I don't think that should be possible, as both files are part of the repository itself, and this should be an atomic operation. But I've added another check that compares the revision numbers anyway; it can also do this for the Stratum 0, if an additional flag is used (disabled by default, as the Stratum 0 will often be behind a firewall).